src/Common/tests/System/Collections/ICollection.Generic.Tests.cs

```diff
@@ -0,0 +1,622 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace System.Collections.Tests
+{
+    /// <summary>
+    /// Contains tests that ensure the correctness of any class that implements the generic
+    /// ICollection interface
+    /// </summary>
+    public abstract class ICollection_Generic_Tests<T> : IEnumerable_Generic_Tests<T>
+    {
+        #region ICollection<T> Helper Methods
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected abstract ICollection<T> GenericICollectionFactory();
+
+        /// <summary>
+        /// Creates an instance of an ICollection{T} that can be used for testing.
+        /// </summary>
+        /// <param name="count">The number of unique items that the returned ICollection{T} contains.</param>
+        /// <returns>An instance of an ICollection{T} that can be used for testing.</returns>
+        protected virtual ICollection<T> GenericICollectionFactory(int count)
+        {
+            ICollection<T> collection = GenericICollectionFactory();
+            AddToCollection(collection, count);
+            return collection;
+        }
+
+        protected virtual bool DuplicateValuesAllowed { get { return true; } }
+        protected virtual bool DefaultValueWhenNotAllowed_Throws { get { return true; } }
+        protected virtual bool IsReadOnly { get { return false; } }
+        protected virtual bool DefaultValueAllowed { get { return true; } }
+        protected virtual IEnumerable<T> InvalidValues { get { return Array.Empty<T>(); } }
+        protected virtual void AddToCollection(ICollection<T> collection, int numberOfItemsToAdd)
+        {
+            int seed = 9600;
+            IEqualityComparer<T> comparer = GetIEqualityComparer();
+            while (collection.Count < numberOfItemsToAdd)
+            {
+                T toAdd = TFactory(seed++);
+                while (collection.Contains(toAdd, comparer) || InvalidValues.Contains(toAdd, comparer))
+                    toAdd = TFactory(seed++);
+                collection.Add(toAdd);
+            }
+        }
+        
+        #endregion
+
+        #region IEnumerable<T> Helper Methods
+
+        protected override IEnumerable<T> GenericIEnumerableFactory(int count)
+        {
+            return GenericICollectionFactory(count);
+        }
+
+        protected override int WaysToModify { get { return 3; } } // add, remove, clear
+
+        protected override void ModifyEnumerable(IEnumerable<T> enumerable, int enumerationCode)
```

Comment:
I haven't looked over it much, but you could potentially do it as a list of `Action<T>`s (via MemberData, probably with some sort of label), which would allow you to specify all "modifying" actions for a given type.